sched/scheduler/jobRunner.go

```diff
@@ -0,0 +1,267 @@
+	jr := jobRunner{
+		job:        job,
+		saga:       saga,
+		sagaState:  sagaState,
+		dist:       dist.NewPoolDistributor(cm.StaticClusterFactory(nodes)),
+		updateChan: make(chan stateUpdate),
+	}
+
+	go jr.updateSagaState()
+	return jr
+}
+
+// Runs the Job associated with this JobRunner to completion
+func (jr jobRunner) runJob() {
+	// don't re-run an already completed saga
+	if jr.sagaState.IsSagaCompleted() {
+		return
+	}
+
+	// if the saga has not been aborted, just run each task in the saga
+	// TODO: this can be made smarter to run the next Best Task instead of just
+	// the next one in order etc....
+	if !jr.sagaState.IsSagaAborted() {
+		for _, task := range jr.job.Tasks {
+
+			if !jr.sagaState.IsTaskCompleted(task.Id) {
+				node := jr.dist.ReserveNode()
+
+				jr.updateChan <- stateUpdate{
+					msgType: saga.StartTask,
+					taskId:  task.Id,
+				}
+
+				jr.wg.Add(1)
+				go func(node cm.Node, task sched.Task) {
+
+					// TODO: After a number of attempts we should stop
+					// Trying to run a task, could be a poison pill
+					// Implement deadletter queue
+					taskExecuted := false
+					for !taskExecuted {
+						err := node.SendMessage(task)
+						if err == nil {
+							taskExecuted = true
+						}
+					}
+
+					jr.updateChan <- stateUpdate{
+						msgType: saga.EndTask,
+						taskId:  task.Id,
+					}
+					jr.dist.ReleaseNode(node)
+					jr.wg.Done()
+				}(node, task)
+			}
+		}
+	}
+
+	if jr.sagaState.IsSagaAborted() {
+		// TODO: we don't have a way to specify comp tasks yet
+		// Once we do they should be ran here.  Currently the
+		// scheduler only supports ForwardRecovery in Sagas so Panic!
+		panic("Rollback Recovery Not Supported Yet!")
+	}
+
+	// wait for all tasks to complete
+	jr.wg.Wait()
+	jr.updateChan <- stateUpdate{
+		msgType: saga.EndSaga,
+	}
+
+	return
+}
+
+type stateUpdate struct {
+	msgType saga.SagaMessageType
+	taskId  string
+	data    []byte
+}
+
+// handles all the sagaState updates
+// TODO: lots of shared retry code, need either Lambdas(go support?) or some refactor to re-use shared code.
+func (jr jobRunner) updateSagaState() {
+
+	for update := range jr.updateChan {
+		switch update.msgType {
+		case saga.StartTask:
+
+			attempts := 0
+			state, err := jr.saga.StartTask(jr.sagaState, update.taskId, update.data)
+
+			// if the attempt to update didn't succeed retry
+			for err != nil && retryableErr(err) && attempts < MAX_RETRY {
+				state, err = jr.saga.StartTask(jr.sagaState, update.taskId, update.data)
+				attempts++
+			}
+
+			// if state transition was successful update internal state
+			if state != nil {
+				jr.sagaState = state
+
+				// state transition was not successful maxed out retries
+				// TODO: for now just panic eventually implement dead letter queue.
+			} else {
+				panic(fmt.Sprintf("Failed to succeesfully log StartTask. sagaId: %v, taskId: %v", jr.sagaState.SagaId(), update.taskId))
```

Comment:
We don't actually want to panic do we? If we exit the loop then runJob could fill up updateChan and then block forever. We'll want a different mechanism I think.